File: src/test/java/de/pflugradts/passbird/application/boot/launcher/PassbirdLauncherTest.java
package de.pflugradts.passbird.application.boot.launcher;

import com.google.inject.Injector;
import com.google.inject.Module;
import de.pflugradts.passbird.application.UserInterfaceAdapterPort;
import de.pflugradts.passbird.application.boot.Bootable;
import de.pflugradts.passbird.application.boot.main.ApplicationModule;
import de.pflugradts.passbird.application.boot.main.PassbirdApplication;
import de.pflugradts.passbird.application.boot.setup.PassbirdSetup;
import de.pflugradts.passbird.application.boot.setup.SetupModule;
import de.pflugradts.passbird.application.configuration.ReadableConfiguration;
import de.pflugradts.passbird.application.configuration.ConfigurationFaker;
import de.pflugradts.passbird.application.configuration.Configuration;
import de.pflugradts.passbird.application.util.FileFaker;
import de.pflugradts.passbird.application.util.GuiceInjector;
import de.pflugradts.passbird.application.util.PathFaker;
import de.pflugradts.passbird.application.util.SystemOperation;
import de.pflugradts.passbird.application.util.SystemOperationFaker;
import org.junit.jupiter.api.BeforeEach;
import org.junit.jupiter.api.Test;
import org.junit.jupiter.api.extension.ExtendWith;
import org.mockito.ArgumentCaptor;
import org.mockito.Captor;
import org.mockito.InjectMocks;
import org.mockito.Mock;
import org.mockito.MockitoAnnotations;
import org.mockito.junit.jupiter.MockitoExtension;
import static org.assertj.core.api.Assertions.assertThat;
import static org.mockito.BDDMockito.given;
import static org.mockito.BDDMockito.then;
import static org.mockito.Mockito.mock;

@ExtendWith(MockitoExtension.class)
class PassbirdLauncherTest {

    @Mock
    private Configuration configuration;
    @Mock
    private UserInterfaceAdapterPort userInterfaceAdapterPort;
    @Mock
    private GuiceInjector guiceInjector;
    @Mock
    private SystemOperation systemOperation;
    @InjectMocks
    private PassbirdLauncher passbirdLauncher;

    @Captor
    private ArgumentCaptor<Module> captor;

    @BeforeEach
    void setup() {
        MockitoAnnotations.initMocks(this);
    }

    @Test
    void shouldLaunchMainApplication_IfKeyStoreExists() {
        // given
        final var expectedBootable = mock(PassbirdApplication.class);
        setupKeyStoreFileMock(true);
        setupInjectorWithCapture(expectedBootable);

        // when
        passbirdLauncher.boot();

        // then
        then(expectedBootable).should().boot();
        assertThat(captor.getAllValues()).isNotNull()
                .extracting("class")
                .containsExactly(ApplicationModule.class);
    }

    @Test
    void shouldLaunchSetup_IfKeyStoreDoesNotExist() {
        // given
        final var expectedBootable = mock(PassbirdSetup.class);
        setupKeyStoreFileMock(false);
        setupInjectorWithCapture(expectedBootable);

        // when
        passbirdLauncher.boot();

        // then
        then(expectedBootable).should().boot();
        assertThat(captor.getAllValues()).isNotNull()
                .extracting("class")
                .containsExactly(SetupModule.class);
    }

    @Test
    void shouldLaunchSetup_IfKeyStoreLocationIsNotSet() {
        // given
        final var expectedBootable = mock(PassbirdSetup.class);
        ConfigurationFaker.faker()
                .forInstance(configuration)
                .withKeyStoreLocation("").fake();
        setupInjectorWithCapture(expectedBootable);

        // when
        passbirdLauncher.boot();

        // then
        then(expectedBootable).should().boot();
        assertThat(captor.getAllValues()).isNotNull()
                .extracting("class")
                .containsExactlyInAnyOrder(SetupModule.class);
    }

    private void setupKeyStoreFileMock(boolean keyStoreExists) {
        final var keystoreDirectory = "tmp";
        final var keystoreFile = FileFaker.faker()
                .fakeFile()
                .withExistsProperty(keyStoreExists).fake();
        final var path = PathFaker.faker()
                .fakePath()
                .withFileResolvingToFilename(keystoreFile, ReadableConfiguration.KEYSTORE_FILENAME).fake();
        SystemOperationFaker.faker()
                .forInstance(systemOperation)
                .withPath(keystoreDirectory, path).fake();
        ConfigurationFaker.faker()
                .forInstance(configuration)
                .withKeyStoreLocation(keystoreDirectory).fake();
    }

    private void setupInjectorWithCapture(final Bootable bootable) {
        final var injector = mock(Injector.class);
        given(injector.getInstance(Bootable.class)).willReturn(bootable);
        given(guiceInjector.create(captor.capture())).willReturn(injector);
    }

}

File: src/test/kotlin/de/pflugradts/passbird/application/boot/launcher/PassbirdLauncherTest.kt
package de.pflugradts.passbird.application.boot.launcher

import de.pflugradts.passbird.application.PassbirdRunContext
import de.pflugradts.passbird.application.RunContext
import de.pflugradts.passbird.application.UserInterfaceAdapterPort
import de.pflugradts.passbird.application.configuration.Configuration
import de.pflugradts.passbird.application.configuration.ReadableConfiguration.Companion.KEYSTORE_FILENAME
import de.pflugradts.passbird.application.configuration.fakeConfiguration
import de.pflugradts.passbird.application.process.migration.MigrationRequest
import de.pflugradts.passbird.application.process.migration.PendingMigration
import de.pflugradts.passbird.application.process.migration.PreLaunchMigrationLocator
import de.pflugradts.passbird.application.toDirectory
import de.pflugradts.passbird.application.toFileName
import de.pflugradts.passbird.application.unmockMain
import de.pflugradts.passbird.application.util.SystemOperation
import de.pflugradts.passbird.application.util.fakePath
import de.pflugradts.passbird.application.util.fakeSystemOperation
import de.pflugradts.passbird.domain.model.slot.Slot
import io.mockk.every
import io.mockk.mockk
import io.mockk.verify
import org.junit.jupiter.api.BeforeEach
import org.junit.jupiter.api.Test
import strikt.api.expectThat
import strikt.assertions.containsExactly

class PassbirdLauncherTest {

    private val configuration = mockk<Configuration>(relaxed = true)
    private val preLaunchMigrationLocator = mockk<PreLaunchMigrationLocator>()
    private val runContext: RunContext = PassbirdRunContext("/tmp".toDirectory(), Slot.DEFAULT)
    private val userInterfaceAdapterPort = mockk<UserInterfaceAdapterPort>(relaxed = true)
    private val systemOperation = mockk<SystemOperation>(relaxed = true)
    private val bootedRoutes = mutableListOf<String>()
    private val passbirdLauncher = PassbirdLauncher(
        configuration,
        preLaunchMigrationLocator,
        runContext,
        userInterfaceAdapterPort,
        systemOperation,
        setupBoot = { bootedRoutes.add("setup") },
        migrationBoot = { _, _ -> bootedRoutes.add("migration") },
        applicationBoot = { bootedRoutes.add("application") },
    )

    @BeforeEach
    fun setup() {
        bootedRoutes.clear()
        every { preLaunchMigrationLocator.detect() } returns MigrationRequest.empty()
    }

    @Test
    fun `should launch main application if key store exists`() {
        // given
        val keyStoreDirectoryName = "/tmp"
        val keyStoreFileName = KEYSTORE_FILENAME
        fakeConfiguration(instance = configuration, withKeyStoreLocation = keyStoreDirectoryName, withAnsiEscapeCodesEnabled = true)
        fakeSystemOperation(
            instance = systemOperation,
            withDirectoryResolvingToFileName = Triple(
                keyStoreDirectoryName.toDirectory(),
                keyStoreFileName.toFileName(),
                fakePath(),
            ),
        )
        every { systemOperation.exists(any<java.nio.file.Path>()) } returns true

        // when
        passbirdLauncher.boot()

        // then
        expectThat(bootedRoutes).containsExactly("application")
    }

    @Test
    fun `should launch migration if migration is required`() {
        // given
        val keyStoreDirectoryName = "/tmp"
        val keyStoreFileName = KEYSTORE_FILENAME
        val keyStoreFilePath = fakePath()
        fakeConfiguration(instance = configuration, withKeyStoreLocation = keyStoreDirectoryName)
        fakeSystemOperation(
            instance = systemOperation,
            withDirectoryResolvingToFileName = Triple(
                keyStoreDirectoryName.toDirectory(),
                keyStoreFileName.toFileName(),
                keyStoreFilePath,
            ),
        )
        every { preLaunchMigrationLocator.detect() } returns MigrationRequest(setOf(PendingMigration("keystore-format")))
        every { systemOperation.exists(keyStoreFilePath) } returns true

        // when
        passbirdLauncher.boot()

        // then
        expectThat(bootedRoutes).containsExactly("migration")
    }

    @Test
    fun `should launch setup if key store does not exist`() {
        // given
        val keyStoreDirectoryName = "/tmp"
        val keyStoreFileName = KEYSTORE_FILENAME
        fakeConfiguration(instance = configuration, withKeyStoreLocation = keyStoreDirectoryName)
        val keyStoreFilePath = fakePath()
        fakeSystemOperation(
            instance = systemOperation,
            withDirectoryResolvingToFileName = Triple(
                keyStoreDirectoryName.toDirectory(),
                keyStoreFileName.toFileName(),
                keyStoreFilePath,
            ),
        )
        every { systemOperation.exists(keyStoreFilePath) } returns false

        // when
        passbirdLauncher.boot()

        // then
        expectThat(bootedRoutes).containsExactly("setup")
        verify(exactly = 0) { preLaunchMigrationLocator.detect() }
    }

    @Test
    fun `should launch setup if key store location is not set`() {
        // given
        val keyStoreDirectoryName = ""
        fakeConfiguration(instance = configuration, withKeyStoreLocation = keyStoreDirectoryName)

        // when
        passbirdLauncher.boot()

        // then
        expectThat(bootedRoutes).containsExactly("setup")
        verify(exactly = 0) { preLaunchMigrationLocator.detect() }
    }
}
